File: backend/app/schemas/signature.py
"""Schemas Pydantic para firmas.

Referencia: docs/API_SPEC.md §3
"""
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict


# ===== Request =====
class SignatureRequest(BaseModel):
    signature_base64: str
    hash_algorithm: str = "SHA-256"
    signature_algorithm: str = "RSA-PSS"
    certificate_pem: str
    certificate_fingerprint: str


# ===== Responses =====
class SignResponse(BaseModel):
    signature_id: int
    document_id: str
    signer_id: int
    status: str
    signed_at: str
    tsa_timestamp: Optional[str] = None
    tsa_authority: Optional[str] = None
    message: str = "Document signed successfully"


class SignerDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    cert_subject: Optional[dict] = None


class SignatureItem(BaseModel):
    id: int
    order: Optional[int] = None
    signer: SignerDetail
    signed_at: datetime
    signature_algorithm: Optional[str] = None
    hash_algorithm: Optional[str] = None
    tsa_timestamp: Optional[datetime] = None
    tsa_authority: Optional[str] = None
    status: str


class SignaturesListResponse(BaseModel):
    document_id: str
    total_signatures: int
    signatures: List[SignatureItem]


class VerifyDetails(BaseModel):
    certificate_valid: bool
    signature_algorithm_valid: bool
    timestamp_valid: bool
    tsa_trusted: bool


class VerifyResponse(BaseModel):
    signature_id: int
    valid: bool
    details: VerifyDetails

File: frontend/src/types/signature.ts
import type { CertificateSubject } from "./certificate";

export type SignatureStatus = "pending" | "signed" | "rejected";

export interface SignatureRequest {
  signature_base64: string;
  hash_algorithm: string;
  signature_algorithm: string;
  certificate_pem: string;
  certificate_fingerprint: string;
}

export interface SignerDetail {
  id: number;
  first_name?: string;
  last_name?: string;
  cert_subject?: CertificateSubject;
}

export interface SignatureItem {
  id: number;
  order?: number;
  signer: SignerDetail;
  signed_at: string;
  signature_algorithm?: string;
  hash_algorithm?: string;
  tsa_timestamp?: string;
  tsa_authority?: string;
  status: SignatureStatus;
}

export interface SignResponse {
  signature_id: number;
  document_id: string;
  signer_id: number;
  status: string;
  signed_at: string;
  tsa_timestamp?: string;
  tsa_authority?: string;
  message: string;
}

export interface SignatureVerificationResult {
  signature_id: number;
  valid: boolean;
  details: {
    certificate_valid: boolean;
    signature_algorithm_valid: boolean;
    timestamp_valid: boolean;
    tsa_trusted: boolean;
  };
}
